contrib/raftexample/raft.go

```diff
@@ -73,23 +87,128 @@ func newRaftNode(id int, peers []string, join bool, proposeC <-chan string,
 		confChangeC: confChangeC,
 		commitC:     commitC,
 		errorC:      errorC,
+		snapC:       snapC,
+		snapDataC:   snapDataC,
 		id:          id,
 		peers:       peers,
 		join:        join,
 		waldir:      fmt.Sprintf("raftexample-%d", id),
+		snapdir:     fmt.Sprintf("raftexample-%d-snap", id),
 		raftStorage: raft.NewMemoryStorage(),
+		snapCount:   defaultSnapCount,
 		stopc:       make(chan struct{}),
 		httpstopc:   make(chan struct{}),
 		httpdonec:   make(chan struct{}),
 		// rest of structure populated after WAL replay
 	}
 	go rc.startRaft()
-	return commitC, errorC
+	return commitC, errorC, rc.snapdir
+}
+
+func (rc *raftNode) saveSnap(snap raftpb.Snapshot) error {
+	walSnap := walpb.Snapshot{
+		Index: snap.Metadata.Index,
+		Term:  snap.Metadata.Term,
+	}
+	if err := rc.wal.SaveSnapshot(walSnap); err != nil {
+		return err
+	}
+	if err := rc.snapshotter.SaveSnap(snap); err != nil {
+		return err
+	}
+	return rc.wal.ReleaseLockTo(snap.Metadata.Index)
+}
+
+func (rc *raftNode) applySnapshot(snapshotToSave raftpb.Snapshot) {
+	if raft.IsEmptySnap(snapshotToSave) {
+		return
+	}
+
+	log.Printf("applying snapshot at index %d", rc.snapshotIndex)
+	defer log.Printf("finished applying snapshot at index %d", rc.snapshotIndex)
+
+	if snapshotToSave.Metadata.Index <= rc.appliedIndex {
+		log.Fatalf("snapshot index [%d] should > progress.appliedIndex [%d] + 1", snapshotToSave.Metadata.Index, rc.appliedIndex)
+	}
+
+	dbFilePath, err := rc.snapshotter.DBFilePath(snapshotToSave.Metadata.Index)
+	if err != nil {
+		panic(err)
+	}
+	fpath := filepath.Join(rc.snapdir, "db")
+	if err := os.Rename(dbFilePath, fpath); err != nil {
+		panic(err)
+	}
+	rc.commitC <- nil // trigger kvstore to load snapshot
+
+	rc.confState = snapshotToSave.Metadata.ConfState
+	rc.snapshotIndex = snapshotToSave.Metadata.Index
+	rc.appliedIndex = snapshotToSave.Metadata.Index
+}
+
+var snapshotCatchUpEntriesN uint64 = 10000
+
+func (rc *raftNode) getSnapshot() []byte {
+	rc.snapC <- struct{}{}
+	return <-rc.snapDataC
```

Comment:
@heyitsanthony Not having `kvstore` field in `raftNode` makes me to add these two fields.
Any other suggestion to make kvstore to store snapshot to disk? `nil` channel to `commitC` would be confusing to do this.

Any feedback?

Thanks.